python/lsst/daf/butler/core/dimensions/_records.py

```diff
@@ -75,6 +78,61 @@ def _subclassDimensionRecord(definition: DimensionElement) -> Type[DimensionReco
     return type(definition.name + ".RecordClass", (DimensionRecord,), d)
 
 
+class SpecificSerializedDimensionRecord(BaseModel, extra="allow"):
+    """Base class for a specific serialized record content.
+
+    Will be updated to specific subclass when attached to
+    `SerializedDimensionRecord`.
+    """
+
+
+def _createSimpleRecordSubclass(definition: DimensionElement) -> Type[SpecificSerializedDimensionRecord]:
+    from ._schema import DimensionElementFields
+    fields = DimensionElementFields(definition)
+    members = {}
+    for field in fields.standard:
+        field_type = field.getPythonType()
+        if field.nullable:
+            field_type = Optional[field_type]  # type: ignore
+        members[field.name] = (field_type, ...)
+    if definition.temporal:
+        members["timespan"] = (Tuple[int, int], ...)  # type: ignore
+    if definition.spatial:
+        members["region"] = (str, ...)
+
+    # mypy does not seem to like create_model
+    return create_model(f"SpecificSerializedDimensionRecord{definition.name.capitalize()}",
+                        __base__=SpecificSerializedDimensionRecord, **members)  # type: ignore
+
+
+class SerializedDimensionRecord(BaseModel):
+    """Simplified model for serializing a `DimensionRecord`."""
+
+    definition: str
+    universe: int
+    record: SpecificSerializedDimensionRecord
+
+    @validator("record")
+    def cast_record(cls, v, values) -> SpecificSerializedDimensionRecord:  # type: ignore # noqa: N805
+        """Cast the generic v to specific v based on the definition."""
+        from ._universe import DimensionUniverse
+
+        if isinstance(v, SpecificSerializedDimensionRecord):
+            if type(v) is not SpecificSerializedDimensionRecord:
+                # If already a subclass, assume the right type.
+                return v
+            # Convert to specific subclass.
+            v = v.dict()
+
+        if not isinstance(v, collections.abc.Mapping):
+            raise ValueError(f"record can either be SpecificDimensionRecord or dict, not {v!r}")
+
+        universe = DimensionUniverse._instances[values["universe"]]
```

Comment:
@TallJimbo I managed to get pydantic to create specific subclasses of serialized records so that validation could be enabled. The main problem being that you need a Universe to get the relevant DimensionElement.  This works because I store the universe version in the model and can therefore get the universe from the singleton. This can potentially significantly simplify the general `from_simple` methods since we could embed the universe version in all the models and they would ask Universe for that version (I assume that other versions should always fail because those versions aren't the one stored in the registry table). Thoughts:

* Currently I'm assuming that we will pass the version number from the client to the server in a parameters model. The server will compare that version to its own version and then if they match would pass that universe to `from_simple` as now.
* Embedding universe version potentially leads to the universe version number appearing at every level of the serialization and in one client/server interaction it should be the same everywhere.
* If we aren't embedding version numbers how could the dimension record be associated with the right universe? Since DimensionUniverse is already a singleton could we have an API that declares the active universe? (maybe as a context manager). One issue is that the model is instantiated by fastapi before we see it so I think I might need to create a universe  when the server starts up since otherwise there might not be one active if the records are being sent from the client (which means a context manager wouldn't work either).
* Currently the version number of a universe is a private attribute. There needs to be a public API for it.
* How would we distinguish a SPHEREx universe version from a Rubin universe version?

Reply:
I'm not actually sure the extra automatic validation is worth it in this case.  I figured having separate classes for different dimensions was useful (as opposed to `SimpleNamespace` or `dict`) because there typos are a real problem and tab-completion of attributes is nice, and I was worried (paranoid?) about using a class with `__slots__` instead of `__dict__` storage for memory efficiency.  But I'm not sure that's really won out over the complexities of having dynamically-defined types even in the pure local `SqlRegistry` case.

And while we may not be able to rely on all clients giving the server consistent dimensions and want to do some validation there, surely there's some pydantic hook that lets you e.g. validate that the keys of a dict match some expectation?  (And if not, I imagine we could always explicitly check in the server code).

I'm already pretty uncomfortable with using the singleton dict keyed on version to work around the pickle limitations in `mulitprocessing` (if that let me use a custom `Unpickler`, I absolutely would, and drop the singleton entirely), as I think it dangerously conflates the kind of version number we want for schema management with something that's really an ugly implementation detail.  It'd be safer to use some kind of digest/hash instead of an incrementing number, at least in the singleton, and maybe elsewhere, at least for everything other than linear, "dimensions used by the Rubin product" human-meaningful schema versioning (which already has problems with e.g. SPHEREx, as you note).

Reply:
> surely there's some pydantic hook that lets you e.g. validate that the keys of a dict match some expectation

I think all validation requires a universe and the pydantic models don't have a universe (if they did everything would work anyhow). I think we'd have to put the validation directly in `from_simple` since that does have a universe. I will switch to that approach.

Reply:
I retained the dynamic pydantic model creation and now create a specialist class to validate everything in `from_simple`. It works well and if I send odd data types or extra fields it does now complain. An obvious problem is that I dynamically create a model every time I call `from_simple` on a record and that seems less than ideal. Caching it works but the hash for DimensionElement is the name of the element so that doesn't seem safe because the universe can change -- should the DimensionElement know the universe it came from?

Reply:
`DimensionElement.universe` does exist, so you could cache on that plus the name (if `DimensionUniverse` is hashable; I'm not sure about that).

I wouldn't add the unvierse to `DimensionElement.__hash__` because that's allowed to be equal for non-equal objects anyway; adding a universe comparison to `DimensionElement.__eq__` might be okay, but there's also all of the `Named*` containers that assume that name equality is equal to object equality (ooh, that might be a problem for non-unique DatasetType names...), so I'd prefer to just define the equality comparison as "only meaningful within a universe" and add the universe explicitly only in contexts (like that cache) where necessary.

Reply:
Thanks. This seems to work (including the universe).